dynawo/sources/Solvers/SolverSIM/DYNSolverSIM.cpp

```diff
@@ -476,88 +355,65 @@ void SolverSIM::handleMaximumTries(int& counter) {
-   * Returns y and yp updated values
-   */
+int
+SolverSIM::callAlgebraicSolver() {
   int flag = 0;
-  if (!skipNextNR_ && model_->sizeY() != 0)
-    flag = callSolverKINEuler();
+  if (skipNextNR_) {
+    return KIN_INITIAL_GUESS_OK;
+  } else {
+    // Step initialization
+    solverKINEuler_->setInitialValues(tSolve_, h_, vYy_);
+
+    // Forcing the Jacobian calculation for the next Newton-Raphson resolution
+    bool noInitSetup = true;
+    if (stats_.nst_ == 0 || factorizationForced_)
+      noInitSetup = false;
+
+    // Call the solving method in Backward Euler method (Newton-Raphson resolution)
+    flag = solverKINEuler_->solve(noInitSetup, skipAlgebraicResidualsEvaluation_);
+
+    // Update statistics
+    updateStatistics();
+  }
+
+  return flag;
+}
 
+SolverSIM::SolverStatus_t SolverSIM::analyzeResult(int& flag) {
   // Analyze the return value and do further treatments if necessary
   if (flag < 0) {
     stats_.ncfn_++;
+    skipNextNR_ = false;
     return NON_CONV;
-  } else {
-    // Evaluate root values after the time step (using updated y and yp)
-    model_->evalG(tSolve_ + h_, g1_);
-    ++stats_.nge_;
-
-    // No root change -> no discrete variable change neither mode change
-    if (std::equal(g0_.begin(), g0_.end(), g1_.begin())) {
-      if (flag == KIN_INITIAL_GUESS_OK) {
-        skipNextNR_ = skipNRIfInitialGuessOK_;
-        if (skipNextNR_)
-          Trace::info() << DYNLog(SolverSIMInitGuessOK) << Trace::endline;
-      }
-      return CONV;
-    } else {
-      // A root change has occurred - Dealing with propagation and algebraic mode detection
-#ifdef _DEBUG_
-      printUnstableRoot(tSolve_ + h_, g0_, g1_);
-#endif
-      /* Save the new values of the root in g0 for comparison after the evalZMode
-       * Calculate the propagation of discrete variable value changes and mode changes
-       */
-      g0_.assign(g1_.begin(), g1_.end());
-      evalZMode(g0_, g1_, tSolve_ + h_);
-
-      if (getState().getFlags(ModeChange)) {
-        skipNextNR_ = false;
-        if (model_->getModeChangeType() != DIFFERENTIAL_MODE)
-          return ROOT_ALG;
-        else if (getState().getFlags(NotSilentZChange)
-            || getState().getFlags(SilentZNotUsedInDiscreteEqChange))
-          return ROOT;
-        return CONV;
-      } else if (getState().getFlags(NotSilentZChange) || getState().getFlags(SilentZNotUsedInDiscreteEqChange)) {  // Z change
-        skipNextNR_ = false;
-        return ROOT;
-      } else {  // Root change without any z or mode change
-        if (flag == KIN_INITIAL_GUESS_OK) {
-          skipNextNR_ = skipNRIfInitialGuessOK_;
-          if (skipNextNR_)
-            Trace::info() << DYNLog(SolverSIMInitGuessOK) << Trace::endline;
-        }
-        return CONV;
-      }
-    }
+  } else if (flag == KIN_INITIAL_GUESS_OK && !skipNextNR_) {
+    skipNextNR_ = skipNRIfInitialGuessOK_;
+    Trace::info() << DYNLog(SolverSIMInitGuessOK) << Trace::endline;
```

Comment:
Should be displayed only if skipNextNR_ is true

Reply:
Moreover we used to have this activated when a root was raised without mode or z change, I don't see it in the new implem?

Reply:
Indeed for the log, I haven't kept in mind that the use of this could be disabled by the user through the par file.

For the root case, I modified a bit the logic to have it clearer (at least from my point of view):
- the fact to skip the next NR is always first set by the algebraic part in the analyzeResult. If one call is ok  with the initial guess, we set skipNextNR_ to true.
- then it is modified if on the discrete variable or mode part, there is a change that will modify the continuous equations, which will cancel the fact that we previously reached steady-state. As such, the skipNextNR_ is only modified to false if modeChange, NotSilentZ or SilentZNotUsedInDiscreteEqChange change. In case of a root with no effect, then skipNextNR_ is not modified by this method and if it has been previously set to true, remains at true.